opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -622,23 +630,30 @@ def start(
 
     def end(self, end_time: Optional[int] = None) -> None:
         with self._lock:
-            if not self.is_recording():
-                return
-            if self.start_time is None:
-                raise RuntimeError("Calling end() on a not started span.")
+            is_recording = self.is_recording()
             has_ended = self.end_time is not None
-            if not has_ended:
-                if self.status is None:
-                    self.status = Status(canonical_code=StatusCanonicalCode.OK)
+            if is_recording:
+                if self.start_time is None:
+                    raise RuntimeError("Calling end() on a not started span.")
+                if not has_ended:
+                    if self.status is None:
+                        self.status = Status(
+                            canonical_code=StatusCanonicalCode.OK
+                        )
 
-                self._end_time = (
-                    end_time if end_time is not None else time_ns()
-                )
+                    self._end_time = (
+                        end_time if end_time is not None else time_ns()
+                    )
+
+                    self._is_recording = False
 
         if has_ended:
             logger.warning("Calling end() on an ended span.")
             return
 
+        if not is_recording:
```

Comment:
Why move this down here? Having the `if not is_recording` check here and the `self._is_recording = false` right above is kind of awkward.

Reply:
Agree readability is not ideal. Per prior comment this was moved down to allow for the has_ended statement to actually be reached and the logger warning thrown if the span has already ended.